### PR TITLE
Add a new layer type: WebGLPointsLayer

### PR DESCRIPTION
**Pull request by jahow**
This is a work-in-progress.

A new  `WebGLPointsLayer` class have been added, which takes only a vector source and a literal style object:

```js
new WebGLPointsLayer({
    source: vectorSource,
    literalStyle: {
      symbol: {
        size: 4,
        color: '#3388FF',
        rotateWithView: false,
        offset: [0, 0],
        opacity: 1
      }
  }
  });
```

Also, the `WebGLPointsRenderer` have been reworked to now use pre-generated shaders and read attributes from the features themselves. This is how the renderer should be used now:

```js
new WebGLPointsLayerRenderer(layer, {
  attributes: [
    {
      name: 'size',
      callback: function(feature) {
        // compute something with the feature
      }
    },
    {
      name: 'weight',
      callback: function(feature) {
        // compute something with the feature
      }
    },
  ],
  vertexShader:
    // shader using attribute a_weight and a_size
  fragmentShader:
    // shader using varying v_weight and v_size
```

The renderer now uses much smaller vertex attribute buffers, and does not make assumption about these attributes (like opacity, color, size etc.). All the render parameters of the points should be handles by user defined attributes.

Things left to do:
- [x] fix hit detection
- [x] restore the possibility to use textures
- [ ] ~allow binding to feature attributes in the literal style format~
- [ ] ~offer predefined symbol shapes (circle, triangle, star...)~

### CODE REVIEW

**Comment by jahow** [comment]
I'm putting this PR for review, otherwise it'll get too large. The new layer type can handle simple style with literal values as well as images.

I'll do the attribute binding in another PR.

Here is the hosted example for the new layer type: https://2104-4723248-gh.circle-artifacts.com/0/examples/webgl-points-layer.html

A sample style for testing image icons:
```json
{
 "symbol": {
  "size": [
   18,
   30
  ],
  "color": "lightyellow",
  "rotateWithView": false,
  "offset": [
   0,
   15
  ],
  "opacity": 1,
  "symbolType": "image",
  "src": "data/icon.png"
 }
}

File: doc/errors/index.md [hunk 1]
```diff
@@ -240,3 +240,11 @@ Support for the `OES_element_index_uint` WebGL extension is mandatory for WebGL
 ### 64
 
 Layer opacity must be a number.
+
+### 65
+
+A symbol literal representation must be defined on the style supplied to a `WebGLPointsLayer` instance.
+
+### 65
```
Comment (fredj) [inline doc/errors/index.md]: duplicate error number

Comment (jahow) [reply doc/errors/index.md]: Thanks, fixed

File: src/ol/style/LiteralStyle.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+};
+
+
+/**
+ * @typedef {Object} LiteralSymbolStyle
+ * @property {number|Array.<number, number>} size Size, mandatory.
+ * @property {SymbolType} symbolType Symbol type to use, either a regular shape or an image.
+ * @property {string} [src] Path to the image to be used for the symbol. Only required with `symbolType: 'image'`.
+ * @property {import("../color.js").Color|string} [color='#FFFFFF'] Color used for the representation (either fill, line or symbol).
+ * @property {number} [opacity=0] Opacity.
```
Comment (fredj) [inline src/ol/style/LiteralStyle.js]:
```suggestion
 * @property {number} [opacity=1] Opacity.
```

File: src/ol/layer/WebGLPoints.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+   */
+  constructor(options) {
+    const baseOptions = assign({}, options);
+
+    super(baseOptions);
+
+    /**
+     * @type {import('../style/LiteralStyle.js').LiteralStyle}
+     */
+    this.literalStyle = options.literalStyle;
```
[the remark this answers is not in the document]
Comment (jahow) [reply src/ol/layer/WebGLPoints.js]: done!

File: examples/icon-sprite-webgl.js [hunk 4]
```diff
@@ -21,47 +21,153 @@ const texture = new Image();
...
-        return coords[component];
+      ],
+      uniforms: {
+        u_texture: texture
       },
-      sizeCallback: function() {
-        return 16;
-      }
+      vertexShader: [
+        'precision mediump float;',
```
[the remark this answers is not in the document]
Comment (jahow) [reply examples/icon-sprite-webgl.js]: as discussed, string literals appear broken in the final output of the example builder.

File: doc/errors/index.md [hunk 5]
```diff
@@ -240,3 +240,11 @@ Support for the `OES_element_index_uint` WebGL extension is mandatory for WebGL
...
 
 Layer opacity must be a number.
+
+### 65
+
+A symbol literal representation must be defined on the style supplied to a `WebGLPointsLayer` instance.
+
+### 66
+
+`forEachFeatureAtCoordinate` cannot be used on a WebGL layer if the hit detection logic has not been enabled.
```
[the remark this answers is not in the document]
Comment (jahow) [reply doc/errors/index.md]: Done

File: src/ol/style/LiteralStyle.js (hunk 2 above)
Comment (jahow) [reply src/ol/style/LiteralStyle.js]: Thanks, fixed

**Comment by jahow** [comment]
Thanks for the review!

**Pull request merged**